### ISSUE TITLE
PDF with utf8mb4/emojis crushes on saving

### ISSUE THREAD

**Issue by DarthLegiON** [mpdf/mpdf#727]
Hello!
I have a problem with mPDF 7.0.0. I create document via HTML fragments and WriteHTML method and where document is closing error fires.
![image](https://user-images.githubusercontent.com/11365194/39134731-fcf6055c-471f-11e8-90e8-7925838816b8.png)

When there is no emoji characters in document - everything is OK. I use default encoding options in Mpdf class, ('UTF-8' as I assume). How can I fix this problem?

PHP Version: 7.1.16
MPDF version 7.0.0

**Comment by finwe** [comment]
Please, post your PHP/HTML code snippet causing this in plain text/code markup. Without it, we are unable to review and/or fix this correctly.

**Comment by DarthLegiON** [comment]
@finwe I've created code example as simpliest as I could and it still causes bug.
```php
$pdf = new Mpdf([
            'tempDir' => Yii::getAlias('@runtime/temp'), // returns output directory
            'format' => 'A4',
        ]);

        $pdf->SetCreator('Example');
        $pdf->SetAuthor('Example');
        $pdf->SetTitle('Document Example');

        $pdf->SetAutoPageBreak(true);
        $pdf->AddPage();
        $pdf->WriteHTML("<p>String With Smiles😊</p>");

        $pdf->Output('example.pdf', 'S');
```
![image](https://user-images.githubusercontent.com/11365194/40475534-8b63f898-5f4a-11e8-814f-039b9252be3c.png)

**Comment by DarthLegiON** [comment]
@finwe it was on 7.0.0. There is another problem on 7.1.0 version: Smiles don't correctly show.
![image](https://user-images.githubusercontent.com/11365194/40476453-b90be5f6-5f4c-11e8-8deb-deeb7605c934.png)

**Comment by marclaporte** [comment]
"For emojis to be displayed, the font has to support them. None of the fonts bundled with mPDF supports emoji characters. You'd have to use a custom font and/or custom font substitutions."

A good use case for packages: https://github.com/mpdf/mpdf/issues/552

**Comment by drsassafras** [comment]
I ran a few tests using Tiki & mPDF v7.1.5 and the emoji all appeared just fine. The example PDF provided earlier does not work, however. So maybe the error is not with mPDF but somewhere else.

On most operating Systems, emoji are not displayed via the font. They are a system function. So the OS replaces the Unicode emoji with its own system-specific emoji. This happens in almost all operating systems and phones. I think Linux might be an exception.

Now emoji are constantly evolving, and they do so with the Unicode standard. So, for example, iOS version x will only have emoji support for up to Unicode x. Were at version 11 now and they create a new version every year. Each year they add about 50 or so emoji. So If you haven't updated your OS within the last year, your not going to be able to see all the emoji out there.

There are packages that will take Unicode and transform it into HTML (with images) but in that case, your emoji image library needs to stay on top of the new emoji updates, otherwise the users OS will have better support for emoji than your package does and it will actually break compatibility while adding complexity!

The last note I wanted to add, is that bundling fonts with emoji support typically won't work. Coloured fonts are all system specific, so if you bundled an emoji font that worked well with OSX, it won't work on Windows, Android or iOS. Yes, even iOS fonts differ from OSX fonts. 

The exception would be if you bundled a font with monochrome emoji (that nobody wants) or if all of your users were on one OS, which is typically not what happens with a PHP project.

So in conclusion, I think the emoji support in mPDF is optimal. I vote for re-closing this bug report with an invalid marker instead of "not supported"
